File: 5_AdvancedAlgorithmsAndComplexity/week3_NPCompleteness/3_budget_allocation/BudgetAllocation.cs
﻿using System.Collections.Generic;
using System.Linq;
using System;

namespace _3_budget_allocation
{
    public static class BudgetAllocation
    {
        private static int[] ToBinary(int value, int digits)
        {
            var result = new int[digits];
            while (value > 0)
            {
                result[(int)Math.Log((value ^ (value - 1)) + 1, 2) - 1] = 1;
                value &= (value - 1);
            }
            return result;
        }

        private static List<int[]> SatClauses()
        {
            return new List<int[]>() { new int[2] { 1, 1 }, new int[3] { 1, -1, 0 } };
        }

        private static List<int[]> UnsatClauses()
        {
            return new List<int[]>() { new int[2] { 2, 1 }, new int[2] { 1, 0 }, new int[2] { -1, 0 } };
        }

        public static string[] Solve(string[] input)
        {
            var nm = input[0].Split(' ').Select(x => int.Parse(x)).ToArray();
            var n = nm[0];
            var m = nm[1];

            if (n < 1 || m < 1)
            {
                return new string[0];
            }

            var A = new int[n][];
            for (int i = 1; i <= n; i++)
            {
                A[i - 1] = input[i].Split(' ').Select(x => int.Parse(x)).ToArray();
            }
            var b = input[n + 1].Split(' ').Select(x => int.Parse(x)).ToArray();

            var output = new List<int[]>() { new int[2] { 0, m } };

            for (int i = 0; i < n; i++)
            {
                var coeffs = new List<int>();
                for (int j = 0; j < m; j++)
                {
                    if (A[i][j] != 0) coeffs.Add(j);
                }
                if (coeffs.Count == 0 && b[i] < 0)
                {
                    output = UnsatClauses();
                    break;
                }
                //check any possible assignment of 3 vars
                var clauseCount = 0;
                for (int k = 0; k < Math.Pow(2, coeffs.Count); k++)
                {
                    var binary = ToBinary(k, coeffs.Count);
                    var sum = 0;
                    for (int l = 0; l < binary.Length; l++)
                    {
                        sum += A[i][coeffs[l]] * binary[l];
                    }
                    if (sum > b[i])
                    {
                        // inequality falsified, add clause
                        var clause = new int[binary.Length + 1];
                        clauseCount++;
                        if (clauseCount == Math.Pow(2, coeffs.Count))
                        {
                            output = UnsatClauses();
                            break;
                        }
                        for (var y = 0; y < binary.Length; y++)
                        {
                            clause[y] = (binary[y] > 0 ? -1 : 1) * (coeffs[y] + 1);
                        }
                        output.Add(clause);
                    }
                }
            }

            output[0][0] = output.Count - 1;
            if (output.Count == 1)
            {
                output = SatClauses();
            }

            return output.Select(x => string.Join(" ", x)).ToArray();
        }

        static void Main(string[] args)
        {
            var nmline = Console.ReadLine();
            var n = int.Parse(nmline.Split(' ')[0]);
            var input = new string[n + 2];
            input[0] = nmline;
            for (int i = 1; i <= n; i++)
            {
                input[i] = Console.ReadLine();
            }
            input[n + 1] = Console.ReadLine();
            foreach (var resline in Solve(input))
            {
                Console.WriteLine(resline);
            }
        }
    }
}

File: 5_AdvancedAlgorithmsAndComplexity/week3_NPCompleteness/3_budget_java/BudgetAllocation.java
import java.io.*;
import java.util.*;

public class BudgetAllocation {
    private final InputReader reader;
    private final OutputWriter writer;

    public BudgetAllocation(InputReader reader, OutputWriter writer) {
        this.reader = reader;
        this.writer = writer;
    }

    public static void main(String[] args) {
        InputReader reader = new InputReader(System.in);
        OutputWriter writer = new OutputWriter(System.out);
        new BudgetAllocation(reader, writer).run();
        writer.writer.flush();
    }

    class ConvertILPToSat {
        int[][] A;
        int[] b;
        ArrayList<int[]> output;

        ConvertILPToSat(int n, int m) {
            A = new int[n][m];
            b = new int[n];
        }

        void printSat() {
            writer.printf("1 1\n");
            writer.printf("1 -1 0\n");
        }

        void printUnsat() {
            writer.printf("2 1\n");
            writer.printf("1 0\n");
            writer.printf("-1 0\n");
        }

        void printOutput() {
            for (int i = 0; i < output.size(); i++) {
                int[] clause = output.get(i);
                String[] stringVars = new String[clause.length];
                for (int j = 0; j < clause.length; j++)
                    stringVars[j] = Integer.toString(clause[j]);
                writer.printf(String.join(" ", stringVars) + "\n");
            }
        }
    }

    private static int[] ToBinary(int value, int digits) {
        int[] result = new int[digits];
        while (value > 0) {
            result[(int) Math.floor(Math.log10((value ^ (value - 1)) + 1) / Math.log10(2)) - 1] = 1;
            value &= (value - 1);
        }
        return result;
    }

    public void run() {
        int n = reader.nextInt();
        int m = reader.nextInt();

        ConvertILPToSat converter = new ConvertILPToSat(n, m);
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                converter.A[i][j] = reader.nextInt();
            }
        }
        for (int i = 0; i < n; ++i) {
            converter.b[i] = reader.nextInt();
        }

        converter.output = new ArrayList<int[]>();
        converter.output.add(new int[] { 0, m });

        for (int i = 0; i < n; i++) {
            List<Integer> coeffs = new ArrayList<Integer>();
            for (int j = 0; j < m; j++) {
                if (converter.A[i][j] != 0)
                    coeffs.add(j);
            }
            if (coeffs.size() == 0 && converter.b[i] < 0) {
                converter.printUnsat();
                return;
            }
            // check any possible assignment of 3 vars
            int clauseCount = 0;
            for (int k = 0; k < Math.pow(2, coeffs.size()); k++) {
                int[] binary = ToBinary(k, coeffs.size());
                int sum = 0;
                for (int l = 0; l < binary.length; l++) {
                    sum += converter.A[i][coeffs.get(l)] * binary[l];
                }
                if (sum > converter.b[i]) {
                    // inequality falsified, add clause
                    int[] clause = new int[binary.length + 1];
                    clauseCount++;
                    if (clauseCount == Math.pow(2, coeffs.size())) {
                        converter.printUnsat();
                        return;
                    }
                    for (int y = 0; y < binary.length; y++) {
                        clause[y] = (binary[y] > 0 ? -1 : 1) * (coeffs.get(y) + 1);
                    }
                    converter.output.add(clause);
                }
            }
        }

        converter.output.get(0)[0] = converter.output.size() - 1;
        if (converter.output.size() == 1) {
            converter.printSat();
        } else {
            converter.printOutput();
        }
    }

    static class InputReader {
        public BufferedReader reader;
        public StringTokenizer tokenizer;

        public InputReader(InputStream stream) {
            reader = new BufferedReader(new InputStreamReader(stream), 32768);
            tokenizer = null;
        }

        public String next() {
            while (tokenizer == null || !tokenizer.hasMoreTokens()) {
                try {
                    tokenizer = new StringTokenizer(reader.readLine());
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }
            return tokenizer.nextToken();
        }

        public int nextInt() {
            return Integer.parseInt(next());
        }
    }

    static class OutputWriter {
        public PrintWriter writer;

        OutputWriter(OutputStream stream) {
            writer = new PrintWriter(stream);
        }

        public void printf(String format, Object... args) {
            writer.print(String.format(Locale.ENGLISH, format, args));
        }
    }
}
